index.rst

```diff
@@ -48,9 +48,118 @@
+
+A year or more ago we introduced a ``NO_SUDO`` flag environment variable
+that can be passed into the container, which bypasses all of the sudo
+steps to escalate privilege.
+
+This comes with one major caveat: it only works if there is a
+pre-provisioned user home directory.  This is fine at NCSA (which is
+where the rule was intended for use) since user home space is
+provisioned as part of account creation and NCSA forces all RSP users to
+have NCSA accounts; it is not fine in a GKE environment using GitHub as
+its authentication back-end.  We have recently proven that everything
+still works as expected (assuming a pre-provisioned filesystem) if the
+container is run with ``allowPrivilegeEscalation: False`` in its
+Kubernetes ``securityContext``.
+
+Recently, in order to allow loading the Multus CNI into containers in
+Rubin-managed Kubernetes environments, necessary for control of
+instruments on multicast networks with SAL, Josh Hoblitt introduced a
+container that must be run as a privileged initContainer inside the
+user's Lab pod before starting the Lab.  This change is already present
+in the mainline Nublado environment.
+
+A Fuller Solution
+=================
+
+Since we've worked out how to run privileged initContainers, all we
+would need to do is to move the provisioning steps into a second
+initContainer, and invoke it (or not) at user pod creation time.  We
+could either restrict its privilege still further by doing the same set
+of semi-privileged user plus sudoers rules we currently do, or just let
+the initContainer run as root, which would simplify the logic within
+considerably.
+
+Then, in all cases, we could invoke the Lab pod directly with the
+correct userid and set of groups, having provisioned those in ConfigMaps
+to mount into ``/etc/passwd`` and ``/etc/group`` (``NO_SUDO`` already
+does this), with the only piece that still requires privilege being the
+creation of user filesystems.
+
+This will be an initContainer that mounts our NFS share volume with
+``no_root_squash`` and effectively runs as root while doing so in order
+to create the user spaces.  That's a little scary, but no scarier than
+what we are doing today, particularly if we restrict the provisionator's
+actions with sudoers rules.
+
+Future Steps
+============
+
+At some point we very likely will to want to replace this mechanism with
+an actual JupyterHub-callable service that does the requested filesystem
+provisioning.  The good thing about the proposed model is that, when we
+do develop the service, then we just stop adding the privileged
+provisioning container to ``initContainers`` at user pod startup, and
+there's nothing else to do.  We cannot, with just the technique
+suggested here, set a cluster-wide ``PodSecurityPolicy``, since every
+user pod may have a privileged component.  However, the choice of
+implementation of the Multus CNI may have tied our hands with respect to
+this in any event.
+
+Estimated Effort
+================
```

Comment:
Could you add estimated effort for writing the service? I think a service that did the work by spawning the same container you were planning on running as an init container would make the most sense (that way the web service itself doesn't have to run as root).

Reply:
Yeah, I can try.  I didn't have a design in my head for writing that, is the thing.  Like, I understand and know what the initContainer entails and I haven't thought through this one.  So basically you're talking a web service, that authenticates incoming requests somehow, that has the power to spawn a privileged container in its namespace, which in turn can attach NFS as root and write some stuff?